Add Neighbors functions

richardartoul opened the pull request:
Reference implementations:

1) https://github.com/corsc/go-geohash/blob/master/geohash_int.go
2) https://github.com/gansidui/geohash/blob/master/geohash.go
3) https://github.com/yinqiwen/geohash-int/blob/master/geohash.c
4) https://github.com/sunng87/node-geohash/blob/master/main.js

Discussion points:

1) I thought this algorithm would be broken around the poles / equator / prime-meridian, but I think your encode function just handles that stuff gracefully? I'm a little confused why points near those areas work fine, but when I plotted the neighbors on a map it looked like it was doing the right thing.

2) Test cases all copied from here: https://github.com/sunng87/node-geohash/blob/master/tests/test.js (Any idea where to get better reference tests?)

3) I'm not confident this is the fastest implementation. I've seen other implementation that seem to be using different algorithms, but this is the only algorithm I could understand.

richardartoul commented:
Ok I auto-generated some tests. Copy paste of the comment I left in the auto-genned test file:

# Test cases generated with ngeohash - Note that I had to remove some of the test cases because they varied slightly (like by really really tiny amounts) from the results generated by this
library. I did some digging and the results were always so close (even if you plot on a Map you can't really tell the difference) that I couldn't comeup with an explanation other than the fact that Javascript uses 64-bit floating point numbers for everything while Golang has support for true integers. If you look through the ngeohash code you'll see that they do many operations on floats that could result in tiny losses of precision.

mmcloughlin commented:
Yes, I have observed that for the other geohash tests. I believe it is a result of both floating point precision, and a different method of mapping lat/lng values to integers. In this library we just use integer rounding (https://github.com/mmcloughlin/geohash/blob/master/geohash.go#L137-L141), but the description of the algorithm describes a binary-search-like technique (https://en.wikipedia.org/wiki/Geohash). I have no idea why they choose to describe it this way, it seems far more complicated to me. But I can see how even though both methods are theoretically equivalent, the combination with floating point precision issues could lead to subtle differences in low bits.

* That said, it does subtly bother me that the file of test cases is not reproducible. What do you think? Is there a way we can avoid this problem, perhaps by limiting ourselves to lower precision?

* Also note that the other generated test cases writes its own source code into the output https://github.com/mmcloughlin/geohash/blob/master/testcases_test.go#L15. This is related to making the file generation reproducible.

mmcloughlin commented on neighbors_testcases_test.go:
```diff
@@ -0,0 +1,514 @@
+//  floating point numbers for everything while Golang has support for true
+//  uint64's. If you look through the ngeohash code you'll see that they do
+//  many operations on floats that could result in tiny losses of precision.
+// Note that I had to remove
```
Repeat?

richardartoul replied:
Will fix.

mmcloughlin commented on neighbors_testcases_test.go:
```diff
@@ -0,0 +1,514 @@
+//   // 2 --> 52. NGeohash has a max precision of 52 (because Javascript)
+//   let intBitdepth = Math.floor(Math.random() * 52) + 2
+//   // ngeohash library specifies that bitdepth must be even
+//   if (intBitdepth %2 != 0) {
```
```javascript
let intBitdepth = 2 * Math.floor(Math.random() * 26) + 2
```

richardartoul replied:
Oh nice thats clever.

mmcloughlin commented on neighbors_testcases_test.go:
```diff
@@ -0,0 +1,514 @@
+package geohash
+
+//	# Test cases generated with ngeohash - Note that I had to remove some of the
```
Indent? As mentioned in other comment, perhaps have your node code write its own source code into the output file?

richardartoul replied:
Will fix.

mmcloughlin commented on geohash_test.go:
```diff
@@ -131,3 +131,60 @@ func TestLeadingZero(t *testing.T) {
+		for i, neighbor := range neighbors {
+			expected := c.hashStrNeighbors[i]
+			if neighbor != expected {
+				t.Errorf("Actual: %v \n expected: %v\n", neighbors, c.hashIntNeighbors)
```
Very nitpicky but all the other failure messages are lowercase.

richardartoul replied:
Will fix

mmcloughlin commented on neighbors_testcases_test.go:
```diff
@@ -0,0 +1,514 @@
+package geohash
```
Does this output file pass `go fmt`?

richardartoul replied:
I'll verify.

mmcloughlin commented:
We also need to regenerate `README.md` from `README.tmpl`. This was done with https://github.com/robertkrimen/godocdown.

mmcloughlin commented on geohash.go:
```diff
@@ -134,6 +134,67 @@ func DecodeInt(hash uint64) (lat, lng float64) {
+	}
+}
+
+// NeighborsInt returns a slice of uint64's that correspond to the provided hash's
```
`uint64's` -> `uint64s` ?

mmcloughlin commented on geohash.go:
```diff
@@ -134,6 +134,67 @@ func DecodeInt(hash uint64) (lat, lng float64) {
+	return NeighborsIntWithPrecision(hash, 64)
+}
+
+// NeighborsIntWithPrecision returns a slice of uint64's that correspond to the
```
`uint64's` -> `uint64s` ?

richardartoul commented on neighbors_testcases_test.go:
```diff
@@ -0,0 +1,1084 @@
+//	 function getSeededRand(seed) {
+//	   return function randWithSeed() {
+//	 		var x;
+//	     do {
```
Looks weird in the preview but if you pull down the code it looks normal

richardartoul commented:
@mmcloughlin Addressed all your feedback (thanks for the thoughtful comments.) Let me know if there are any other changes you'd like, but I think its starting to look pretty good.
